src/filter/mod.rs

```diff
@@ -11,13 +11,127 @@ use image::{GenericImage, GenericImageView, GrayImage, ImageBuffer, Luma, Pixel,
 use crate::definitions::{Clamp, Image};
 use crate::integral_image::{column_running_sum, row_running_sum};
 use crate::map::{ChannelMap, WithChannel};
-use num::Num;
+use num::{abs, pow, Num};
 
 use crate::math::cast;
 use conv::ValueInto;
 use std::cmp::{max, min};
 use std::f32;
 
+/// Bilateral filtering of grayscale images.
+pub fn bilateral_filter(
+    image: &GrayImage,
+    win_size: u32,
+    sigma_color: f32,
+    sigma_spatial: f32,
+    n_bins: u32,
+) -> Image<Luma<u8>> {
+    /// Un-normalized Gaussian weights for look-up tables.
+    fn guassian_weight(x: f32, sigma_squared: f32) -> f32 {
+        return (-0.5 * x.powi(2) / sigma_squared).exp();
+    }
+
+    /// Effectively a meshgrid command with flattened outputs.
+    fn win_coords(win_size: u32) -> (Vec<i32>, Vec<i32>) {
+        let win_start = (-(win_size as f32) / 2.0).floor() as i32;
+        let win_end = (win_size as f32 / 2.0).floor() as i32 + 1;
+        let win_range = win_start..win_end;
+        let v = win_range.collect::<Vec<i32>>();
+        let cc: Vec<i32> = v
+            .iter()
+            .cycle()
+            .take(v.len() * v.len())
+            .into_iter()
+            .cloned()
+            .collect();
+        let mut rr = Vec::new();
+        let win_range = win_start..win_end;
+        for i in win_range {
+            rr.append(&mut vec![i; (win_size + 1) as usize]);
+        }
+        return (rr, cc);
+    }
+
+    /// Create look-up table of Gaussian weights for color dimension.
+    fn compute_color_lut(bins: u32, sigma: f32, max_value: f32) -> Vec<f32> {
+        let v = (0..bins as i32).collect::<Vec<i32>>();
+        let step_size = max_value / bins as f32;
+        let values = v.iter().map(|&x| x as f32 * step_size).collect::<Vec<_>>();
+        let sigma_squared = sigma.powi(2);
+        let gauss_weights = values
+            .iter()
+            .map(|&x| guassian_weight(x, sigma_squared))
+            .collect::<Vec<_>>();
+        gauss_weights
+    }
+
+    /// Create look-up table of weights corresponding to flattened 2-D Gaussian kernel.
+    fn compute_spatial_lut(win_size: u32, sigma: f32) -> Vec<f32> {
+        let (rr, cc) = win_coords(win_size);
+        let mut gauss_weights = Vec::new();
+        let it = rr.iter().zip(cc.iter());
+        let sigma_squared = sigma.powi(2);
+        for (r, c) in it {
+            let dist = f32::sqrt(pow(*r as f32, 2) + pow(*c as f32, 2));
+            gauss_weights.push(guassian_weight(dist, sigma_squared));
+        }
+        gauss_weights
+    }
+
+    let (n_cols, n_rows) = image.dimensions();
```

Comment:
Nit: these are called 'width' and 'height' in most of the rest of the code in this library. I'd prefer to use those names here for consistency.